kernel/src/utilities/streaming_process_slice.rs

```diff
@@ -0,0 +1,285 @@
+    /// argument from the buffer, do not modify the version, do not change any
+    /// flags that are controlled by the process or otherwise violate the
+    /// protocol, and correctly increment the `offset` value.
+    ///
+    /// - `SIZE`: if the underlying slice is not large enough to fit the
+    ///   flags field and the offset field (5 bytes).
+    fn write_header(&self, header: StreamingProcessSliceHeader) -> Result<(), ErrorCode> {
+        // Write the offset first, to avoid modifying the buffer if it's too
+        // small to fit the offset, but large enough to hold the flags byte:
+        let offset_bytes = u32::to_ne_bytes(header.offset);
+        self.slice
+            .get(Self::RANGE_OFFSET)
+            .ok_or(ErrorCode::SIZE)?
+            .copy_from_slice_or_err(&offset_bytes)
+            .map_err(|_| ErrorCode::SIZE)?;
+
+        let flags_bytes: [u8; 1] = [(StreamingProcessSliceFlags::VERSION
+            .val(header.version as u8)
+            + StreamingProcessSliceFlags::RESERVED::RESERVED0
+            + StreamingProcessSliceFlags::HALT.val(header.halt as u8)
+            + StreamingProcessSliceFlags::EXCEEDED.val(header.exceeded as u8))
+        .value];
+        self.slice
+            .get(Self::RANGE_FLAGS)
+            .ok_or(ErrorCode::SIZE)?
+            .copy_from_slice_or_err(&flags_bytes)
+            .map_err(|_| ErrorCode::SIZE)?;
+
+        Ok(())
+    }
+
+    /// Access the payload data portion of the underlying slice.
+    ///
+    /// This method does not perform any validation of the buffer version or
+    /// data. It must only be used on slices of version 0. If the underlying
+    /// slice is too small to hold the header fields, this will return an empty
+    /// slice.
+    fn payload_slice(&self) -> &WriteableProcessSlice {
+        self.slice
+            .get(Self::RANGE_DATA)
+            .unwrap_or((&mut [][..]).into())
+    }
+
+    /// Append a chunk of data to the slice.
+    ///
+    /// If the underlying slice has a correct `flags` and `offset` value, is not
+    /// halted, and has sufficient space for this `data` chunk, this function
+    /// returns the updated buffer offset (set to one past the last written
+    /// byte).
+    ///
+    /// This function fails with:
+    /// - `INVAL`: if the version is not `0`, or the reserved flags are not
+    ///   cleared.
+    /// - `BUSY`: if both the `halt` and `exceeded` flags are set. In this case,
+    ///   the slice will not be modified.
+    /// - `SIZE`: if the underlying slice is not large enough to fit the
+    ///   flags field and the offset field (5 bytes). In this case, the
+    ///   `exceeded` flag will be set and the slice will not be modified.
+    /// - `FAIL`: would need to increment offset beyond 2**32 - 1. Neither the
+    ///   payload slice nor any header fields will be modified.
+    pub fn append_chunk(&self, chunk: &[u8]) -> Result<u32, ErrorCode> {
```

Comment:
Actually, I think that we may want to return a `bool` here indicating whether this was the first write to this buffer. This would allow a driver to only schedule a single upcall per allowed buffer, as long as the process follows the above contract and always allows new buffers with offset set to `0`.

@bradjc @alexandruradovici What do you think? I'm afraid that if we don't have this, we may run out of space in the upcall queue quickly and drop other upcalls for this process, especially for short, frequent chunks such as produced by a streaming ADC driver.

Reply:
It seems like we want both. The bool for exactly the reason you mention, but the offset in case the capsule wants to signal the upcall and tell the app where to look. Right now the capsule could check the offset == 0, but that may not work for version 2.

Reply:
Actually, given that we don't publically expose the offset before an `append_chunk` operation, the capsule couldn't easily do that today either (well, I guess it could do the math itself by comparing the current chunks length against the offset after append).

But yes, I agree we should have both!